### PR TITLE
Timeline API added and issues in notes and attachment fixed

### PR DESCRIPTION
**Pull request by UmashriR**
Issue in notes and attachment fixed

### CODE REVIEW

File: ZCRMiOS/Classes/CommonUtil.swift [hunk 1]
```diff
@@ -1,680 +0,0 @@
...
-    public static var RESPONSE_ROOT_KEY_NIL = "RESPONSE_ROOT_KEY_NIL"
-    public static var FILE_SIZE_EXCEEDED = "FILE_SIZE_EXCEEDED"
-    public static var MAX_COUNT_EXCEEDED = "MAX_COUNT_EXCEEDED"
-    public static var FIELD_NOT_FOUND = "FIELD_NOT_FOUND"
-    public static var OAUTHTOKEN_NIL = "OAUTHTOKEN_NIL"
-    public static var OAUTH_FETCH_ERROR = "OAUTH_FETCH_ERROR"
-    public static var UNABLE_TO_CONSTRUCT_URL = "UNABLE_TO_CONSTRUCT_URL"
-    public static var INVALID_FILE_TYPE = "INVALID_FILE_TYPE"
-    public static var INVALID_MODULE = "INVALID_MODULE"
-}
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/CommonUtil.swift]: Why did you remove CommonUtil.Swift file?

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 2]
```diff
@@ -317,6 +317,43 @@ internal class EntityAPIHandler : CommonAPIHandler
         }
     }
     
+    internal func getTimelineEvents( page : Int, perPage : Int, filter : String?, completion : @escaping( Result.DataResponse< [ ZCRMTimelineEvent ], BulkAPIResponse > ) -> () )
+    {
+        setJSONRootKey( key : JSONRootKey.TIMELINES )
+        var timelines : [ZCRMTimelineEvent] = [ZCRMTimelineEvent]()
+        setUrlPath(urlPath: "/\(self.recordDelegate.moduleAPIName)/\(self.recordDelegate.recordId)/timelines")
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: check recordDelegate has recordId.

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 3]
```diff
@@ -317,6 +317,43 @@ internal class EntityAPIHandler : CommonAPIHandler
...
+        addRequestParam(param: "page", value: String(page))
+        addRequestParam(param: "per_page", value: String(perPage))
+        let request : APIRequest = APIRequest(handler: self )
+        print( "Request : \( request.toString() )" )
+        
+        request.getBulkAPIResponse { ( resultType ) in
+            do{
+                let bulkResponse = try resultType.resolve()
+                let responseJSON = bulkResponse.getResponseJSON()
+                if responseJSON.isEmpty == false
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: else must throw an error.

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 4]
```diff
@@ -756,7 +793,10 @@ internal class EntityAPIHandler : CommonAPIHandler
             }
             else if( ResponseJSONKeys.participants == fieldAPIName )
             {
-                self.setParticipants( participantsArray : value as! [ [ String : Any ] ] )
+                if recordDetails.hasValue(forKey: ResponseJSONKeys.participants)
+                {
+                    self.setParticipants( participantsArray : value as! [ [ String : Any ] ] )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift:761]: Don't do force unwrap.

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 5]
```diff
@@ -1038,6 +1078,40 @@ internal class EntityAPIHandler : CommonAPIHandler
         participant.isInvited = participantDetails.getBoolean( key : ResponseJSONKeys.invited ) 
         return participant
     }
+    
+    private func getZCRMTimelineEvent( timelineDetails : [ String : Any ] ) throws -> ZCRMTimelineEvent
+    {
+        let record : ZCRMRecordDelegate = ZCRMRecordDelegate( recordId: timelineDetails.getDictionary(key: ResponseJSONKeys.record).getInt64(key: ResponseJSONKeys.id), moduleAPIName: timelineDetails.getDictionary(key: ResponseJSONKeys.record).getDictionary(key: ResponseJSONKeys.module).getString(key: ResponseJSONKeys.name))
+        let timeline : ZCRMTimelineEvent = ZCRMTimelineEvent(action: timelineDetails.getString(key: "action"), record : record)
+        if timelineDetails.hasValue(forKey: "audited_time") == false
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: "audited_time", "action" should be added in ResponseJSON keys

File: ZCRMiOS/Classes/RelatedListAPIHandler.swift [hunk 6]
```diff
@@ -62,9 +62,18 @@ internal class RelatedListAPIHandler : CommonAPIHandler
                         let recordsList:[[String:Any]] = responseJSON.getArrayOfDictionaries( key : self.getJSONRootKey() )
                         for recordDetails in recordsList
                         {
-                            let record : ZCRMRecord = ZCRMRecord(moduleAPIName: relatedList.apiName)
-                            try EntityAPIHandler(record: record).setRecordProperties(recordDetails: recordDetails)
-                            records.append(record)
+                            if relatedList.apiName == "Activities"
+                            {
+                                let record : ZCRMRecord = ZCRMRecord(moduleAPIName: recordDetails.getString(key: "Activity_Type"))
+                                try EntityAPIHandler(record: record).setRecordProperties(recordDetails: recordDetails)
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/RelatedListAPIHandler.swift]: "Activities", "Activity_Type" should be added in ResponseJSON keys.

File: ZCRMiOS/Classes/ZCRMNote.swift [hunk 7]
```diff
@@ -45,6 +45,20 @@ open class ZCRMNote : ZCRMEntity
         self.attachments?.append( attachment )
     }
     
+    fileprivate func removeAttachment( attachmentId : Int64 )
+    {
+        var index : Int = Int()
+        for count in 0..<self.attachments!.count - 1
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMNote.swift]: Don't use force unwrap.

File: ZCRMiOS/Classes/ZCRMNote.swift [hunk 8]
```diff
@@ -83,13 +97,29 @@ open class ZCRMNote : ZCRMEntity
...
         ZCRMModuleRelation(relatedListAPIName: "Attachments", parentModuleAPIName: "Notes").uploadAttachmentWithPath(ofParentRecord: ZCRMRecordDelegate(recordId: self.id, moduleAPIName: "Notes"), filePath: filePath) { ( result ) in
+            do
+            {
+                try self.addAttachment(attachment: result.resolve().data)
+            }
+            catch
+            {
+                completion( .failure( typeCastToZCRMError( error ) ) )
+            }
             completion( result )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMNote.swift]: completion should be in do block.

File: ZCRMiOS/Classes/ZCRMNote.swift [hunk 9]
```diff
@@ -83,13 +97,29 @@ open class ZCRMNote : ZCRMEntity
...
         ZCRMModuleRelation(relatedListAPIName: "Attachments", parentModuleAPIName: "Notes").uploadAttachmentWithData( ofParentRecord: ZCRMRecordDelegate(recordId: self.id, moduleAPIName: "Notes"), fileName : fileName, data : data ) { ( result ) in
+            do
+            {
+                try self.addAttachment(attachment: result.resolve().data)
+            }
+            catch
+            {
+                completion( .failure( typeCastToZCRMError( error ) ) )
+            }
             completion( result )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMNote.swift]: completion should be in the do block.

File: ZCRMiOS/Classes/ZCRMNote.swift [hunk 10]
```diff
@@ -114,6 +144,15 @@ open class ZCRMNote : ZCRMEntity
     public func deleteAttachment( attachmentId : Int64, completion : @escaping( Result.Response< APIResponse > ) -> () )
     {
         ZCRMModuleRelation( relatedListAPIName : "Attachments", parentModuleAPIName : "Notes" ).deleteAttachment( ofParentRecord : ZCRMRecordDelegate(recordId: self.id, moduleAPIName: "Notes"), attachmentId : attachmentId ) { ( result ) in
+            do
+            {
+                let _ = try result.resolve()
+                self.removeAttachment(attachmentId: attachmentId)
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMNote.swift]: add response status check before removing the attachment in attachment array.

File: ZCRMiOS/Classes/ZCRMOrganisationDelegate.swift [hunk 11]
```diff
@@ -19,6 +19,31 @@ open class ZCRMOrganisationDelegate : ZCRMEntity
...
+    {
+        return ZCRMProfile(name:  name)
+    }
+    
+    public func getRoleDelegate(roleId: Int64, roleName: String) -> ZCRMRoleDelegate
+    {
+        return ZCRMRoleDelegate(roleId: roleId, roleName: roleName)
+    }
+    
+    public func newRole(name: String) -> ZCRMRole
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMOrganisationDelegate.swift]: Is it necessary?

File: ZCRMiOS/Classes/ZCRMOrganisationDelegate.swift [hunk 12]
```diff
@@ -19,6 +19,31 @@ open class ZCRMOrganisationDelegate : ZCRMEntity
...
+    {
+        return ZCRMProfileDelegate(profileId: profileId, profileName: profileName)
+    }
+    
+    public func getProfileDelegate(profileId: Int64, profileName: String, isDefault: Bool) -> ZCRMProfileDelegate
+    {
+        return ZCRMProfileDelegate(profileId: profileId, profileName: profileName, isDefault: isDefault)
+    }
+    
+    public func newProfile( name : String ) -> ZCRMProfile
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMOrganisationDelegate.swift]: Is it necessary?

File: ZCRMiOS/Classes/ZCRMStageDelegate.swift [hunk 13]
```diff
@@ -0,0 +1,8 @@
+//
+//  ZCRMStageDelegate.swift
+//  ZCRMiOS
+//
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/ZCRMStageDelegate.swift]: why did you add this file?

File: ZCRMiOS/Classes/EntityAPIHandler.swift [hunk 14]
```diff
@@ -1038,6 +1089,40 @@ internal class EntityAPIHandler : CommonAPIHandler
...
+        let doneByDetails : [ String : Any ] = timelineDetails.getDictionary( key : ResponseJSONKeys.doneBy )
+        let doneBy : ZCRMUserDelegate = ZCRMUserDelegate( id : doneByDetails.getInt64( key : ResponseJSONKeys.id ), name : doneByDetails.getString( key : ResponseJSONKeys.name ) )
+        timeline.doneBy = doneBy
+        if timelineDetails.hasValue(forKey: ResponseJSONKeys.source) == false
+        {
+            throw ZCRMError.InValidError( code : ErrorCode.VALUE_NIL, message : "\( ResponseJSONKeys.source ) is must not be nil" )
+        }
+        timeline.sourceName = timelineDetails.getDictionary(key: ResponseJSONKeys.source).getString(key: ResponseJSONKeys.name)
+        timeline.automationType = timelineDetails.optDictionary(key: ResponseJSONKeys.automationDetails)?.optString(key: ResponseJSONKeys.type)
+        timeline.automationRule = timelineDetails.optDictionary(key: ResponseJSONKeys.automationDetails)?.optDictionary(key: ResponseJSONKeys.rule)?.optString(key: ResponseJSONKeys.name)
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/EntityAPIHandler.swift]: Why do you fetch automation details dictionary two times?

File: ZCRMiOS/Classes/MassEntityAPIHandler.swift [hunk 15]
```diff
@@ -54,14 +54,14 @@ internal class MassEntityAPIHandler : CommonAPIHandler
                         try EntityAPIHandler(record: record).setRecordProperties(recordDetails: recordJSON)
                         updatedRecords.append(record)
                         entityResponse.setData(data: record)
+                        bulkResponse.setData( data : updatedRecords )
```
Comment (BoopathyPSiva) [inline ZCRMiOS/Classes/MassEntityAPIHandler.swift]: Revert this change.

File: ZCRMiOS/Classes/ZCRMModuleDelegate.swift [hunk 16]
```diff
@@ -26,6 +26,11 @@ open class ZCRMModuleDelegate : ZCRMEntity
         return ZCRMRecordDelegate( recordId : id, moduleAPIName : apiName )
     }
     
+    public func getCustomView( cvName : String ) -> ZCRMCustomView
+    {
+        return ZCRMCustomView(cvName: cvName, moduleAPIName: self.apiName)
```
Comment (vijayakrishnap) [inline ZCRMiOS/Classes/ZCRMModuleDelegate.swift]: This should hit request to return fully filled ZCRMCustomView object right?
If something will return an object without hitting request, it should return only delegate object.

**Review by vijayakrishnap — CHANGES_REQUESTED** [review]

**Review by vijayakrishnap — APPROVED** [review]

**Pull request merged**